core/drivers/rstctrl/rstctrl.c

```diff
@@ -0,0 +1,63 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2021, Linaro Limited
+ */
+
+#include <assert.h>
+#include <drivers/rstctrl.h>
+#include <kernel/dt.h>
+#include <kernel/dt_driver.h>
+#include <kernel/spinlock.h>
+#include <libfdt.h>
+#include <stdint.h>
+
+/* Global reset controller access lock */
+
+TEE_Result rstctrl_get_exclusive(struct rstctrl *rstctrl)
+{
+	uint32_t exceptions = 0;
+	TEE_Result res = TEE_ERROR_ACCESS_CONFLICT;
+	static unsigned int rstctrl_lock = SPINLOCK_UNLOCK;
+
+	exceptions = cpu_spin_lock_xsave(&rstctrl_lock);
+
+	if (!rstctrl->exclusive) {
+		rstctrl->exclusive = true;
+		res = TEE_SUCCESS;
+	}
+
+	cpu_spin_unlock_xrestore(&rstctrl_lock, exceptions);
+
+	return res;
+}
+
+void rstctrl_put_exclusive(struct rstctrl *rstctrl)
+{
+	assert(rstctrl->exclusive);
+
+	rstctrl->exclusive = false;
```

Comment:
I would prefer if this was assigned using a `WRITE_ONCE()` macro. We don't have that macro yet, but it should be easy to implement using `__compiler_atomic_store()`